### ISSUE TITLE
MSFT texture and packing extensions DDS requirements

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/glTF#1359]
Could MSFT extensions be more sound about allowed DDS flags/features (i.e., mip levels, pixelformats, depth, old FourCC vs new DXT10 header, channel bitmasks, etc)?

/cc @bghgary @sbtron

**Comment by bghgary** [comment]
We are using this by passing the DDS container directly to DirectX APIs in specific use cases. If there is more interest, perhaps we can define a set of flags/features that can be implemented by more platforms, but I'm not sure how useful this is until we have more use cases.

**Comment by Ziriax** [comment]
@bghgary We are using DDS at work, combined with our [Maya2glTF](https://github.com/WonderMediaProductions/Maya2glTF) exporter, and inhouse DirectX 11 GLTF renderer (we hope to make the latter open source). 

So we will also use the MSFT extension in the future. 

IMHO GLTF is not just for the web, it's nice to finally have an open standard for 3D.

**Comment by bghgary** [comment]
> GLTF is not just for the web

Certainly. We are strong proponents of that. But my main question is if there are use cases where MSFT_dds is useful on a platform that does not have DirectX APIs.

**Comment by Ziriax** [comment]
That is a good question... BabylonJS supports DDS, but has a converter to some custom format, so that doesn't count I guess.

**Comment by bghgary** [comment]
> BabylonJS supports DDS, but has a converter to some custom format

Right, but that's not a scenario for glTF. We don't currently have any plans to add MSFT_texture_dds extension support to BabylonJS glTF loader.

**Comment by zellski** [comment]
As someone who is currently in the middle of writing a backend service that provides a reasonable texture compression API for all major pixel *and* container formats, I have become a little cranky about this very topic.

DDS is just a file format, right? A simple one, too. It has no actual reliance on anything DirectX. It has no other platform-specific dependencies. Anyone can write a complete parser./generator for DDS, for any platform. Why would we not ship ETC1 in a DDS container, or even ASTC (we could keep it simple and just make FourCC identifiers for all the block sizes)?

If we're saying KTX and KTX2 are so fundamentally a better fit to glTF that we should actively discourage its coupling to DDS, that's a valid argument. If we're saying the forthcoming texture transmission extension will make all this irrelevant, that's valid too. But if we're not going that far, if we're going to flirt with DDS at all, then I suggest we ignore cultural and historical correlations between platform and format, and treat it as a just a file format.

**Comment by bghgary** [comment]
I don't think I'm trying to say DDS is not generic and can't be used everywhere. I am saying this extension is currently being used for a specific scenario for a specific platform. Documenting the exact details of this isn't that useful in my opinion. If we have other scenarios, we can and should add more details so that it can be implemented more consistently. We should also make it a multi-vendor extension instead of a vendor extension.

**Comment by lexaknyazev** [comment]
> DDS is just a file format, right? A simple one, too. It has no actual reliance on anything DirectX. It has no other platform-specific dependencies. Anyone can write a complete parser./generator for DDS, for any platform.

DDS is not defined well enough to be considered for generic interoperable usage. A few examples from MSDN:
> When you write .dds files, you should set the DDSD_CAPS and DDSD_PIXELFORMAT flags, and for mipmapped textures you should also set the DDSD_MIPMAPCOUNT flag. However, when you read a .dds file, you should not rely on the DDSD_CAPS, DDSD_PIXELFORMAT, and DDSD_MIPMAPCOUNT flags being set because some writers of such a file might not set these flags.

> When you write .dds files, you should set the DDSCAPS_TEXTURE flag, and for multiple surfaces you should also set the DDSCAPS_COMPLEX flag. However, when you read a .dds file, you should not rely on the DDSCAPS_TEXTURE and DDSCAPS_COMPLEX flags being set because some writers of such a file might not set these flags.

> The D3DX library (for example, D3DX11.lib) and other similar libraries unreliably or inconsistently provide the pitch value in the dwPitchOrLinearSize member of the DDS_HEADER structure. Therefore, when you read and write to DDS files, we recommend that you compute the pitch in one of the following ways for the indicated formats ...

> The pitch value that you calculate does not always equal the pitch that the runtime supplies, which is DWORD-aligned in some situations and byte-aligned in other situations. Therefore, we recommend that you copy a scan line at a time rather than try to copy the whole image in one copy.

---

> Why would we not ship ETC1 in a DDS container, or even ASTC (we could keep it simple and just make FourCC identifiers for all the block sizes)?

Only Microsoft has authority over DDS FourCC registry, we cannot just start using whatever we want. Moreover, FourCC is deprecated and with new DDS header syntax it is always `DX10`, while the actual format is provided by `DXGI_FORMAT` field (which is also defined by MS as a part of DirectX headers).

---

My question about DDS is a bit more simple. I wanted to add some DDS support to glTF-Validator so that assets with DDS textures are recognized properly. For example, if supplied DDS file for base color contains a cubemap instead of a 2D texture, that would be a validation error. However, if such assets are used strictly within Microsoft closed ecosystem and are never exposed to external viewers, there's little need to bother with validation.

**Comment by zellski** [comment]
Thanks for the clarifications. I guess what I’m after is just an EXT_ version of the simple extension, but with an awareness that in the process of getting there we might see other folks want to add to it.

**Comment by donmccurdy** [comment]
Note that (for lack of any alternative) three.js also has basic support for MSFT_texture_dds: https://github.com/mrdoob/three.js/pull/13820. It's currently undocumented and was merged with the warning that we may remove it when an alternative is available. Despite that, I'm not convinced it should be renamed as EXT_... what are the advantages of doing so?

**Comment by donmccurdy** [comment]
I'll open a PR describing how Vendor extensions relate to the glTF format, and the IP implications. That should probably be visible somewhere in the extensions folder, in any case.

As for whether an EXT_ extension should be created prior to the official KHR texture transmission extension, I don't know ... thoughts would be welcome.

**Comment by donmccurdy** [comment]
We are removing support for DDS textures embedded in glTF files in three.js (https://github.com/mrdoob/three.js/pull/21271), in favor of `KHR_texture_basisu`. Standalone DDS textures will continue to be supported. Are there still significant reasons to consider renaming this as EXT_?
